cpp/test/IceSSL/configuration/OpenSSLTests.cpp

```diff
@@ -0,0 +1,531 @@
+    }
+    return initialize(initData);
+}
+
+void
+clientValidatesServerUsingCAFile(Test::TestHelper* helper, const string& testDir)
+{
+    cout << "client validates server using CAFile... " << flush;
+    SSL_CTX* serverSSLContext = SSL_CTX_new(TLS_method());
+
+    const string serverCertFile = testDir + "/../certs/s_rsa_ca1_pub.pem";
+    const string serverKeyFile = testDir + "/../certs/s_rsa_ca1_priv.pem";
+    SSL_CTX_use_certificate_chain_file(serverSSLContext, serverCertFile.c_str());
+    SSL_CTX_use_PrivateKey_file(serverSSLContext, serverKeyFile.c_str(), SSL_FILETYPE_PEM);
+    SSL_CTX_set_default_passwd_cb(serverSSLContext, passwordCallback);
+
+    const string clientCAFile = testDir + "/../certs/cacert1.pem";
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_method());
+    SSL_CTX_load_verify_file(clientSSLContext, clientCAFile.c_str());
+
+    try
+    {
+        Ice::SSL::ServerAuthenticationOptions serverAuthenticationOptions{
+            .serverSSLContextSelectionCallback =
+                [serverSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(serverSSLContext);
+                return serverSSLContext;
+            },
+            .sslNewSessionCallback = [](::SSL* ssl, const string&) { SSL_set_verify(ssl, SSL_VERIFY_NONE, 0); }};
+        Ice::CommunicatorHolder serverCommunicator(createServer(serverAuthenticationOptions, helper));
+
+        Ice::SSL::ClientAuthenticationOptions clientAuthenticationOptions{
+            .clientSSLContextSelectionCallback = [clientSSLContext](const string&)
+            {
+                SSL_CTX_up_ref(clientSSLContext);
+                return clientSSLContext;
+            }};
+        Ice::CommunicatorHolder clientCommunicator(createClient(clientAuthenticationOptions));
+
+        ServerPrx obj(clientCommunicator.communicator(), "server:" + helper->getTestEndpoint(20, "ssl"));
+        obj->ice_ping();
+    }
+    catch (...)
+    {
+        SSL_CTX_free(serverSSLContext);
+        SSL_CTX_free(clientSSLContext);
+        throw;
+    }
+    SSL_CTX_free(serverSSLContext);
+    SSL_CTX_free(clientSSLContext);
+    cout << "ok" << endl;
+}
+
+void
+clientRejectsServerUsingCAFile(Test::TestHelper* helper, const string& testDir)
+{
+    cout << "client rejects server using CAFile... " << flush;
+    SSL_CTX* serverSSLContext = SSL_CTX_new(TLS_server_method());
+
+    const string serverCertFile = testDir + "/../certs/s_rsa_ca1_pub.pem";
+    const string serverKeyFile = testDir + "/../certs/s_rsa_ca1_priv.pem";
+    SSL_CTX_use_certificate_chain_file(serverSSLContext, serverCertFile.c_str());
+    SSL_CTX_use_PrivateKey_file(serverSSLContext, serverKeyFile.c_str(), SSL_FILETYPE_PEM);
+    SSL_CTX_set_default_passwd_cb(serverSSLContext, passwordCallback);
+
+    // const string clientCAFile = testDir + "/../certs/cacert2.pem";
+    SSL_CTX* clientSSLContext = SSL_CTX_new(TLS_client_method());
```

Comment:
Shouldn't we be loading the CA since that's in the test name? I assume this still fails since there's just no CA configured at all.

Reply:
Yes, commented this for debugging and forget about it.